runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt

```diff
@@ -0,0 +1,117 @@
+ *     public data class SubSealedB(val i: Int) : SimpleSealed()
+ * }
+ *
+ * // will perform correct polymorphic serialization and deserialization:
+ * Json.stringify(SimpleSealed.serializer(), SubSealedA("foo"))
+ * ```
+ *
+ * However, it is still possible to register additional subclasses later using regular [SerializersModule].
+ * It is required when one of inheritors of sealed class is an abstract class:
+ *
+ * ```
+ * @Serializable
+ * sealed class ProtocolWithAbstractClass {
+ *     @Serializable
+ *     abstract class Message : ProtocolWithAbstractClass() {
+ *         @Serializable
+ *         data class StringMessage(val description: String, val message: String) : Message()
+ *
+ *         @Serializable
+ *         data class IntMessage(val description: String, val message: Int) : Message()
+ *     }
+ *
+ *     @Serializable
+ *     data class ErrorMessage(val error: String) : ProtocolWithAbstractClass()
+ * }
+ * ```
+ *
+ * In this case, `ErrorMessage` would be registered automatically by plugin,
+ * but `StringMessage` and `IntMessage` require usual registration, as described in [PolymorphicSerializer] documentation:
+ *
+ * ```
+ * val abstractContext = SerializersModule {
+ *     polymorphic(ProtocolWithAbstractClass::class, ProtocolWithAbstractClass.Message::class) {
+ *         ProtocolWithAbstractClass.Message.IntMessage::class with ProtocolWithAbstractClass.Message.IntMessage.serializer()
+ *         ProtocolWithAbstractClass.Message.StringMessage::class with ProtocolWithAbstractClass.Message.StringMessage.serializer()
+ *         // no need to register ProtocolWithAbstractClass.ErrorMessage
+ *     }
+ * }
+ * ```
+ *
+ */
+public class SealedClassSerializer<T : Any>(
+    serialName: String,
+    override val baseClass: KClass<T>,
+    subclasses: Array<KClass<out T>>,
+    subSerializers: Array<KSerializer<out T>>
+) : AbstractPolymorphicSerializer<T>() {
+
+    private val backingMap: Map<KClass<out T>, KSerializer<out T>>
+
+    private val inverseMap: Map<String, KSerializer<out T>>
+
+
+    init {
+        require(subclasses.size == subSerializers.size) { "Arrays of classes and serializers must have the same length" }
+        backingMap = subclasses.zip(subSerializers).toMap()
+        inverseMap = backingMap.values.associateBy { serializer -> serializer.descriptor.name }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun findPolymorphicSerializer(decoder: CompositeDecoder, klassName: String): KSerializer<out T> {
+        return inverseMap[klassName]
+                ?: super.findPolymorphicSerializer(decoder, klassName)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun findPolymorphicSerializer(encoder: Encoder, value: T): KSerializer<out T> {
+        return backingMap[value::class]
+                ?: super.findPolymorphicSerializer(encoder, value)
+    }
+
+    override val descriptor: SerialDescriptor = SealedClassDescriptor(serialName, subSerializers.map { it.descriptor })
```

Comment:
code style, all properties should be declared in a single place (before functions)